### PR TITLE
OSX: Add spl compile work

### PR DESCRIPTION
**Pull request by lundman**
OSX compile start from module/os/macos/Makefile

Ignores userland, just tries to compile some files in SPL kernel.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: config/kernel.m4 [hunk 1]
```diff
@@ -266,6 +269,16 @@ AC_DEFUN([ZFS_AC_KERNEL], [
...
+	AC_ARG_WITH([macos],
+		AS_HELP_STRING([--with-macos=PATH],
+		[Path to macOS source]),
+		[kernelsrc="$withval/sys"])
+
+	AC_ARG_WITH(macos-obj,
+		AS_HELP_STRING([--with-macos-obj=PATH],
+		[Path to macOS build objects]),
+		[kernelbuild="$withval/$kernelsrc"])
+
```
Comment (lundman) [inline config/kernel.m4:281]: Maybe these could live inside `ZFS_AC_KERNEL_SRC_MACOS_HEADERS`

File: config/kernel.m4 [hunk 2]
```diff
@@ -328,7 +342,10 @@ AC_DEFUN([ZFS_AC_KERNEL], [
 		])
 		kernverinc=$kernelsrc
 		kernvervar=__FreeBSD_version
-	], [
+		kernsrcver=__FreeBSD_version
+	])
+
+	AS_IF([test -z $kernsrcver], [
```
Comment (lundman) [inline config/kernel.m4:348]: Wrap this in the `$kernsrcver` check, so it skips it if already set.

File: config/zfs-build.m4 [hunk 3]
```diff
@@ -453,6 +455,7 @@ AC_DEFUN([ZFS_AC_DEFAULT_PACKAGE], [
 		ubuntu)     DEFAULT_PACKAGE=deb  ;;
 		debian)     DEFAULT_PACKAGE=deb  ;;
 		freebsd)    DEFAULT_PACKAGE=pkg  ;;
+		apple)      DEFAULT_PACKAGE=pkg  ;;
```
Comment (lundman) [inline config/zfs-build.m4:458]: This will be a future problem? Both using pkg

File: configure.ac [hunk 4]
```diff
@@ -29,9 +29,9 @@
  *
  * CDDL HEADER END
  */
-
-AC_INIT(m4_esyscmd(grep ^Name: META | cut -d ':' -f 2 | tr -d ' \n'),
-	m4_esyscmd(grep ^Version: META | cut -d ':' -f 2 | tr -d ' \n'))
+m4_define([NAME], m4_esyscmd([awk '/^Name:/ {print $2}' META | tr -d '\n']))
+m4_define([VERS], m4_esyscmd([awk '/^Version:/ {print $2}' META | tr -d '\n']))
+AC_INIT(NAME, VERS)
```
Comment (lundman) [inline configure.ac:34]: oops, will revert - i was fighting autoconf and what you see here is desperation

File: configure.ac [hunk 5]
```diff
@@ -54,6 +54,11 @@ AM_PROG_CC_C_O
 AX_CODE_COVERAGE
 _AM_PROG_TAR(pax)
 
+dnl macOS additionals
+AC_PROG_CXX(clang++)
+AC_PROG_OBJC(clang)
+AC_PROG_OBJCXX(clang++)
+
```
Comment (lundman) [inline configure.ac:61]: Shouldn't matter that these don't exist on other platforms.

File: include/os/macos/spl/sys/Makefile.am [hunk 6]
```diff
@@ -0,0 +1,5 @@
+KERNEL_H = \
+	$(top_srcdir)/include/os/macos/spl/sys/atomic.h \
+	$(top_srcdir)/include/os/macos/spl/sys/condvar.h \
+	$(top_srcdir)/include/os/macos/spl/sys/mutex.h \
+	$(top_srcdir)/include/os/macos/spl/sys/types.h
```
Comment (lundman) [inline include/os/macos/spl/sys/Makefile.am]: Why does linux define `KERNEL_H`? Is it used for something

File: config/zfs-build.m4 (hunk 3 above)
Comment (behlendorf) [reply config/zfs-build.m4:458]: `pkg` was lightly (if at all) used as an alias on Linux for "build the right packages for my system".  If it's useful we could extend that logic to macos and freebsd.

File: include/os/macos/spl/sys/Makefile.am (hunk 6 above)
Comment (behlendorf) [reply include/os/macos/spl/sys/Makefile.am]: The headers were split in to `KERNEL_H`, `USER_H` and `COMMON_H` to allow installing the user and kernel devel headers in the right locations.  On the Linux side there are other kernel modules (Lustre) which need to build against ZFS, and in user space there are a couple libzfs and libzfs_core consumers.

File: module/os/macos/zfs/test.c [hunk 7]
```diff
@@ -0,0 +1,7 @@
+
+roger()
+{
+	spl_start();
+}
```
Comment (lundman) [inline module/os/macos/zfs/test.c:5]: Just here to call something in spl, or the lazy linker leaves everything out.

File: include/os/macos/spl/sys/vnode.h [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+
+// Be aware that Apple defines "typedef struct vnode *vnode_t" and
+// ZFS uses "typedef struct vnode vnode_t".
+#undef uio_t
+#undef vnode_t
+#include_next <sys/vnode.h>
+#define vnode_t struct vnode
+#define uio_t struct uio
+
```
Comment (lundman) [inline include/os/macos/spl/sys/vnode.h:53]: This is probably the ugliest hack we have to do when trying to fit in with xnu kernel.

**Comment by lundman** [comment]
This is probably as far as I will take this. Build environment for spl and producing a kext. The actual zfs porting should probably wait until freebsd has been committed.

**Comment by lundman** [comment]
That.. was surprisingly easy to get to work. (CI compile of macos)

File: configure.ac (hunk 4 above)
Comment (JMoVS) [reply configure.ac:34]: lol

**Pull request closed without merging**